src/thermo/PengRobinson.cpp

```diff
@@ -0,0 +1,817 @@
+        return -1.0;
+    }
+    return mmw / molarVolLast;
+}
+
+double PengRobinson::densSpinodalLiquid() const
+{
+    double Vroot[3];
+    double T = temperature();
+    int nsol = solveCubic(T, pressure(), m_a, m_b, m_aAlpha_mix, Vroot);
+    if (nsol != 3) {
+        return critDensity();
+    }
+
+    auto resid = [this, T](double v) {
+        double pp;
+        return dpdVCalc(T, v, pp);
+    };
+
+    boost::uintmax_t maxiter = 100;
+    std::pair<double, double> vv = bmt::toms748_solve(
+        resid, Vroot[0], Vroot[1], bmt::eps_tolerance<double>(48), maxiter);
+
+    double mmw = meanMolecularWeight();
+    return mmw / (0.5 * (vv.first + vv.second));
+}
+
+double PengRobinson::densSpinodalGas() const
+{
+    double Vroot[3];
+    double T = temperature();
+    int nsol = solveCubic(T, pressure(), m_a, m_b, m_aAlpha_mix, Vroot);
+    if (nsol != 3) {
+        return critDensity();
+    }
+
+    auto resid = [this, T](double v) {
+        double pp;
+        return dpdVCalc(T, v, pp);
+    };
+
+    boost::uintmax_t maxiter = 100;
+    std::pair<double, double> vv = bmt::toms748_solve(
+        resid, Vroot[1], Vroot[2], bmt::eps_tolerance<double>(48), maxiter);
+
+    double mmw = meanMolecularWeight();
+    return mmw / (0.5 * (vv.first + vv.second));
+}
+
+double PengRobinson::dpdVCalc(double T, double molarVol, double& presCalc) const
+{
+    double den = molarVol * molarVol + 2 * molarVol * m_b - m_b * m_b;
+    presCalc = GasConstant * T / (molarVol - m_b) - m_aAlpha_mix / den;
+
+    double vpb = molarVol + m_b;
+    double vmb = molarVol - m_b;
+    double dpdv = -GasConstant * T / (vmb * vmb) + 2 * m_aAlpha_mix * vpb / (den*den);
+    return dpdv;
+}
+
+void PengRobinson::calculatePressureDerivatives() const
+{
+    double T = temperature();
+    double mv = molarVolume();
+    double pres;
+
+    m_dpdV = dpdVCalc(T, mv, pres);
+    double vmb = mv - m_b;
+    double den = mv * mv + 2 * mv * m_b - m_b * m_b;
+    m_dpdT = (GasConstant / vmb - daAlpha_dT() / den);
+}
+
+void PengRobinson::updateMixingExpressions()
+{
+    double temp = temperature();
+
+    // Update individual alpha
+    for (size_t j = 0; j < m_kk; j++) {
+        double critTemp_j = speciesCritTemperature(m_a_coeffs(j,j), m_b_coeffs[j]);
+        double sqt_alpha = 1 + m_kappa[j] * (1 - sqrt(temp / critTemp_j));
+        m_alpha[j] = sqt_alpha*sqt_alpha;
+    }
+
+    //Update aAlpha_i, j
+    for (size_t i = 0; i < m_kk; i++) {
+        for (size_t j = 0; j < m_kk; j++) {
+            m_aAlpha_binary(i, j) = sqrt(m_alpha[i] * m_alpha[j]) * m_a_coeffs(i,j);
+        }
+    }
+    calculateAB(m_a,m_b,m_aAlpha_mix);
+}
+
+void PengRobinson::calculateAB(double& aCalc, double& bCalc, double& aAlphaCalc) const
+{
+    bCalc = 0.0;
+    aCalc = 0.0;
+    aAlphaCalc = 0.0;
+    for (size_t i = 0; i < m_kk; i++) {
+        bCalc += moleFractions_[i] * m_b_coeffs[i];
+        for (size_t j = 0; j < m_kk; j++) {
+            double a_vec_Curr = m_a_coeffs(i, j);
```

Comment:
Is this variable necessary?  Is there any advantage, over just using `m_a_coeffs(i, j)` directly in the next line?

Reply:
Corrected.